### PR TITLE
Implement scalars

### PR DESCRIPTION
**Pull request by kfurgol**
**Description**

Changes proposed in this pull request:

- Implemented scalars
- Added scalar tests 

**Related issue(s)**
#51

### CODE REVIEW

File: components/director/internal/graphql/scalars/timestamp.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+	t, err := time.Parse(time.RFC3339, tmpStr)
+	if err != nil {
+		return err
+	}
+
+	*y = Timestamp(t)
+
+	return nil
+}
```
Comment (crabtree) [inline components/director/internal/graphql/scalars/timestamp.go]: Add a new line here :)

File: components/director/internal/graphql/scalars/helpers.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+func convertToString(in interface{}) (string, error) {
+	if in == nil {
+		return "", errors.New("input should not be nil")
+	}
+
+	value, ok := in.(string)
+	if !ok {
+		return "", errors.Errorf("unexpected input type: %T, should be map[string][]string", in)
```
Comment (crabtree) [inline components/director/internal/graphql/scalars/helpers.go]: This error seems to be a bit misleading here.

File: components/director/internal/graphql/scalars/timestamp.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+type Timestamp time.Time
+
+func (y *Timestamp) UnmarshalGQL(v interface{}) error {
+
+	if v == nil {
+		return nil
+	}
+
+	tmpStr, ok := v.(string)
```
Comment (crabtree) [inline components/director/internal/graphql/scalars/timestamp.go]: Is there any purpose you do not use convertToString here?

File: components/director/internal/graphql/scalars/annotations_test.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+	//when
+	err := a.UnmarshalGQL(fixAnnotations)
+
+	//then
+	require.NoError(t, err)
+	assert.Equal(t, a, expectedAnnotations)
+}
+
+func TestAnnotations_MarshalGQL(t *testing.T) {
```
Comment (crabtree) [inline components/director/internal/graphql/scalars/annotations_test.go]: This smells like two test cases. Consider splitting it.

File: components/director/internal/graphql/scalars/helpers.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
+package scalars
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/pkg/errors"
+)
+
+func writeResponse(in interface{}, w io.Writer) error {
```
Comment (crabtree) [inline components/director/internal/graphql/scalars/helpers.go]: Please rename this method to be more self-describing. This has nothing with writing the response.

File: components/director/internal/graphql/scalars/timestamp.go (hunk 1 above)
Comment (kfurgol) [reply components/director/internal/graphql/scalars/timestamp.go]: done

File: components/director/internal/graphql/scalars/helpers.go (hunk 2 above)
Comment (kfurgol) [reply components/director/internal/graphql/scalars/helpers.go]: done

File: components/director/internal/graphql/scalars/timestamp.go (hunk 3 above)
Comment (kfurgol) [reply components/director/internal/graphql/scalars/timestamp.go]: no, I just missed it! I've changed it

File: components/director/internal/graphql/scalars/annotations_test.go (hunk 4 above)
Comment (kfurgol) [reply components/director/internal/graphql/scalars/annotations_test.go]: done

File: components/director/internal/graphql/scalars/helpers.go (hunk 5 above)
Comment (kfurgol) [reply components/director/internal/graphql/scalars/helpers.go]: I've renamed it to `marshalAndWrite`, if you have a better name please give me a hint

File: components/director/internal/graphql/scalars/annotations_test.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAnnotations_UnmarshalGQL(t *testing.T) {
```
Comment (crabtree) [inline components/director/internal/graphql/scalars/annotations_test.go]: How about covering here case with multiple values, as we have also in Marshalling func.

File: components/director/internal/graphql/scalars/clob.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
+package scalars
+
+import (
+	"io"
+	"log"
+)
+
+type CLOB string
```
Comment (crabtree) [inline components/director/internal/graphql/scalars/clob.go]: I think we should store CLOB as []byte.

File: components/director/internal/graphql/scalars/helpers.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
+package scalars
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/pkg/errors"
+)
+
+func marshalAndWrite(in interface{}, w io.Writer) error {
```
Comment (crabtree) [inline components/director/internal/graphql/scalars/helpers.go]: What about marshallToWriter?

File: components/director/internal/graphql/scalars/clob.go (hunk 7 above)
Comment (pkosiec) [reply components/director/internal/graphql/scalars/clob.go]: Hmm, what is the reason of that?

File: components/director/internal/graphql/scalars/annotations.go [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package scalars
```
Comment (pkosiec) [inline components/director/internal/graphql/scalars/annotations.go]: Scalars shouldn't be in separate package as they are also GraphQL types. Please move them back to `graphql` package, keeping one file per scalar.

File: components/director/internal/graphql/scalars/annotations_test.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+	//when
+	err := a.UnmarshalGQL(fixAnnotations)
+
+	//then
+	require.NoError(t, err)
+	assert.Equal(t, a, expectedAnnotations)
+}
+
+func TestAnnotations_MarshalGQL_Int(t *testing.T) {
```
Comment (pkosiec) [inline components/director/internal/graphql/scalars/annotations_test.go]: I would have single `TestAnnotations` func with two `t.Run` cases inside.

File: components/director/internal/graphql/scalars/helpers.go (hunk 8 above)
Comment (pkosiec) [reply components/director/internal/graphql/scalars/helpers.go]: `marshallToWriter` sounds a little bit confusing.
What about `writeMarshalled` or `writeAsBytes`?

File: components/director/internal/graphql/scalars/helpers.go [hunk 11]
```diff
@@ -0,0 +1,52 @@
+package scalars
```
Comment (pkosiec) [inline components/director/internal/graphql/scalars/helpers.go]: I don't like the idea of `helpers.go` file. Maybe we could move it to `pkg` as exported functions. However, we need a good name for that package 😄  Maybe `pkg/scalar`?

File: components/director/internal/graphql/scalars/labels_test.go [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"bytes"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestLabels_UnmarshalGQL(t *testing.T) {
+	//given
```
Comment (pkosiec) [inline components/director/internal/graphql/scalars/labels_test.go]: Yes, error case could be also good to test for all scalars

File: components/director/internal/graphql/scalars/annotations_test.go (hunk 10 above)
Comment (crabtree) [inline components/director/internal/graphql/scalars/annotations_test.go]: As those two following test functions are similar in the implementation only differs by given arguments you can reuse the implementation by providing the test payload and expectations as a list and iterating over those values.

For example https://tutorialedge.net/golang/improving-your-tests-with-testify-go/#combining-testify-with-table-driven-tests

File: components/director/internal/graphql/scalars/clob.go (hunk 7 above)
Comment (crabtree) [reply components/director/internal/graphql/scalars/clob.go]: For me, it is more natural to store some unstructured, possibly binary data in []byte. You can easily then convert to a string if needed.

File: components/director/internal/graphql/scalars/helpers.go (hunk 8 above)
Comment (crabtree) [inline components/director/internal/graphql/scalars/helpers.go]:
```suggestion
func marshallToWriter(in interface{}, w io.Writer) error {
```

Comment (pkosiec) [reply components/director/internal/graphql/scalars/helpers.go]: Hmm, marshallToWriter is confusing. I had two other ideas: 
`writeMarshalled` or `writeBytes`

File: components/director/internal/graphql/scalars/annotations_test.go (hunk 6 above)
Comment (kfurgol) [reply components/director/internal/graphql/scalars/annotations_test.go]: done

File: components/director/internal/graphql/scalars/annotations.go (hunk 9 above)
Comment (kfurgol) [reply components/director/internal/graphql/scalars/annotations.go]: done

File: components/director/internal/graphql/scalars/annotations_test.go (hunk 10 above)
Comment (kfurgol) [reply components/director/internal/graphql/scalars/annotations_test.go]: Did it with use of Table-Driven Tests

Comment (kfurgol) [reply components/director/internal/graphql/scalars/annotations_test.go]: done

File: components/director/internal/graphql/scalars/clob.go (hunk 7 above)
Comment (kfurgol) [reply components/director/internal/graphql/scalars/clob.go]: changed to []byte

File: components/director/internal/graphql/scalars/helpers.go (hunk 11 above)
Comment (kfurgol) [reply components/director/internal/graphql/scalars/helpers.go]: moved to pkg/scalar/

File: components/director/internal/graphql/scalars/helpers.go (hunk 8 above)
Comment (kfurgol) [reply components/director/internal/graphql/scalars/helpers.go]: changed to `writeMarshalled`

File: components/director/internal/graphql/scalars/labels_test.go (hunk 12 above)
Comment (kfurgol) [reply components/director/internal/graphql/scalars/labels_test.go]: done

File: components/director/pkg/scalar/helpers.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
+package scalar
```
Comment (pkosiec) [inline components/director/pkg/scalar/helpers.go]: Rename the file to scalar.go

File: components/director/internal/graphql/tenant.go [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+	*y = Tenant(val)
+
+	return nil
+}
+
+func (y Tenant) MarshalGQL(w io.Writer) {
+	_, err := w.Write([]byte(y))
+	if err != nil {
+		log.Printf("error with writing %T", y)
```
Comment (pkosiec) [inline components/director/internal/graphql/tenant.go]: As @crabtree noticed, this is quite bad to log an error. But of course for now we can't return error. I reported an issue for gqlgen: https://github.com/99designs/gqlgen/issues/743

**Pull request merged**